dm/worker/worker_test.go

```diff
@@ -170,3 +174,223 @@ func (t *testServer) TestTaskAutoResume(c *C) {
 		return false
 	}), IsTrue)
 }
+
+type testWorkerEtcdCompact struct{}
+
+var _ = Suite(&testWorkerEtcdCompact{})
+
+func (t *testWorkerEtcdCompact) SetUpSuite(c *C) {
+	NewRelayHolder = NewDummyRelayHolder
+	NewSubTask = func(cfg *config.SubTaskConfig, etcdClient *clientv3.Client) *SubTask {
+		cfg.UseRelay = false
+		return NewRealSubTask(cfg, etcdClient)
+	}
+	createUnits = func(cfg *config.SubTaskConfig, etcdClient *clientv3.Client) []unit.Unit {
+		mockDumper := NewMockUnit(pb.UnitType_Dump)
+		mockLoader := NewMockUnit(pb.UnitType_Load)
+		mockSync := NewMockUnit(pb.UnitType_Sync)
+		return []unit.Unit{mockDumper, mockLoader, mockSync}
+	}
+}
+
+func (t *testWorkerEtcdCompact) TearDownSuite(c *C) {
+	NewRelayHolder = NewRealRelayHolder
+	NewSubTask = NewRealSubTask
+	createUnits = createRealUnits
+}
+
+func (t *testWorkerEtcdCompact) TestWatchSubtaskStageEtcdCompact(c *C) {
+	var (
+		masterAddr   = "127.0.0.1:8291"
```

Comment:
How about using a random port (like using `github.com/pingcap/pd/pkg/tempurl`)?
If so, ditto for L314.